Allow unimposition from the "Split pages" menu

selfhoster1312 opened the pull request:
This is a much requested feature by my friends who make zines, and was also requested in #1067. This is the only feature from booklet-imposer missing in pdfarranger (booklet-imposer is deprecated).

This PR adds an unimposition checkbox in the split pages menu.

TODO:

- [x] work on page selection, not just entire document
- [x] make sure page selection is contiguous
- [x] update translations
- [ ] tests pass

I wanted to have feedback to help finalize this PR so i'm posting early. Here's a preview:

![unimpose](https://github.com/user-attachments/assets/0edec8eb-427c-41b8-a6eb-06ccebbd07fa)

kbengs commented:
Imo it would fit better in a "Booklet" menu with two submenus "Impose booklet" and "De-impose booklet"
ie "generate booklet" is renamed to "Impose booklet".

> make sure page selection is contiguous

It would be possible to grey out "De-impose booklet" when selection is not continuous.

When doing model changes you should use `with self.render_lock():` I'll look more at the code later.

dreua commented:
With only a first glance at the code I'd say the "split_pages" function is getting way to long. Would it be feasible to split (haha) into at least two functions?

kbengs commented:
Unimpose is fine for me. I am unfamiliar with all the suggested words. :smiley: I suggest the names should make it clear that one is the opposite of the other.

selfhoster1312 commented:
Does this look better? Gathering new feedback before trying to write tests and updating translations...

There's now a Booklet menu containing:

- Generate (imposition)
- Split (unimposition)

I've moved the `_crops` splitter method out of the `Dialog` class so i can reuse it without magic tricks, maybe we could add some tests here too... I've added an actual error dialog when the selection is not contiguous.

Maybe we should make sure page size is the same in selection?

selfhoster1312 commented:
So i can't run tests myself but otherwise i think this is ready for review. Some notes:

- the translation/testing process is very badly explained in docs (but i will open dedicated issue for this)
- it's strange that imposition adds blank pages even for 2 pages (which could be a simple A4 pane that you'd like to duplicate to print 2 times the same A5 leaflet for each A4 sheet), and maybe stranger that it adds those white pages at the end? i would think adding blank pages on page 1/n-1 would make most sense so you keep the cover intact
- i'm really not sure about the end of the split_booklet method, i just copied that from split_pages, and it looks very different from generate_booklet method end... i don't know what i'm doing, i don't know why i would use the exporter, or not... please pay attention to that part :)

Otherwise i think it's looking good?

kbengs commented:
Some comments:
* Generate and Split sound good to me
* Mnemonics is missing ie _Generate and _Split in menu.ui (to allow use of Alt + G or S)
*  Split need to be greyed out when nothing is selected, otherwise there are errors.
* the end of split_booklet looks good. The comment can be removed. (end of generate_booklet has clear_selected() which does the same lines)

I'll look more later.

selfhoster1312 commented:
I rebased and addressed your comments. I think you can review :)

kbengs reviewed (commented):
Sorry for commenting again. I had this written before but forgot to push on "submit review" Also if you can clean the history (force push) so that you do not have so many commits. (ie not have commits where you fix your own code) Maybe one commit for the code changes, one for the test and one for the translation is enough.

selfhoster1312 commented:
Applied your suggestions and rebased on main. I believe the test may still fail.

If you know how to fix the test feel free to push on my branch directly! Since i cannot run the tests locally it's not very convenient for me...

kbengs commented on tests/test.py:
```diff
@@ -729,7 +729,19 @@ def test_05_buggy_exif(self):
 
-    def test_06_quit(self):
+    def test_06_split_booklet(self):
+        self.test_O3_booklet()
```
You probably just need to remove this line self.test_O3_booklet()

kbengs replied:
or change O to 0..

kbengs commented:
I think its not possible to force push to your branch so I could not make the tests in one commit. Could you fixup the two test commits into one?
